### ISSUE TITLE
Use create functions signatures and use typing for them

### ISSUE THREAD

**Issue by KOLANICH** [pythonnet/pythonnet#705]
Now functions are quite prototypeless. They use `*args` and `**kwargs` even though C# is strongly typed.

This is not good.

Since signatures are known we can create prototypes easily using `ast` and a (pretty outdated, I have revorked it a lot, but haven't updated the draft, but the essentialls are the same) example can be found on https://github.com/jpype-project/jpype/issues/297 .

**Comment by denfromufa** [comment]
Is this something already supported in IronPython with .NET interop or in comtypes/pywin32 with COM interop?

**Comment by Thrameos** [comment]
I had a working prototype for dealing overloads for JPype for JProxy.  The method is place `@Override` on each of the methods in the class and then a '@Implements` on the class declaration.  The `@Override` allows the method to be mutated to add the required signatures and the `@Implements` collects the list of overriden methods and creates a master stub in the parent class.  However, this depends on the use of a module global table as you can not get access to the class under construction in python until after the class construction is finished, thus overloads must be stored in a table until they can be collected.  I am not sure about this projects issues, but a similar solution may be possible.

**Comment by filmor** [comment]
I'm not sure this is worth the hassle. Python's type hint system does not support (to my knowledge) multiple "full" signatures, instead one would have to fake all arguments that may take different types as `Union`s, additionally explicit handling for default parameters would need to be added.

**Comment by MHDante** [comment]
I actually have a complete working example available at https://github.com/MHDante/pythonnet-stub-generator

It produces fully functional typings for pretty much all of PythonNet (even the weird cases)

I would love to work with someone to refactor this, and add testing and documentation.

If you'd like to try it, you can install it using 

```
dotnet tool install --global pythonnetstubgenerator.tool
GeneratePythonNetStubs --dest-path="../py_project/typings" --target-dlls="dll_folder/MyLib1.dll;other_folder/MyLib2.dll"
```

**Comment by KOLANICH** [comment]
I guess it may make sense to rewrite that into Python.

**Comment by MHDante** [comment]
Yes, my implementation creates overload signatures of __call__ for proxy objects.
